registry/remote/repository.go

```diff
@@ -194,27 +194,43 @@ func (r *Repository) FetchReference(ctx context.Context, reference string) (ocis
 // the same base reference with the Repository r, ParseReference returns a
 // wrapped error ErrInvalidReference.
 func (r *Repository) ParseReference(reference string) (registry.Reference, error) {
-	ref, err := registry.ParseReference(reference)
-	if err != nil {
-		// reference is not a FQDN
-		if index := strings.IndexByte(reference, '@'); index != -1 {
-			// drop tag since the digest is present
-			reference = reference[index+1:]
-		}
+	var err error
+	var ref registry.Reference
+
+	if ref, err = registry.ParseReference(reference); err != nil {
 		ref = registry.Reference{
 			Registry:   r.Reference.Registry,
 			Repository: r.Reference.Repository,
 			Reference:  reference,
 		}
-		if err = ref.ValidateReference(); err != nil {
+
+		// reference is not a FQDN
+		if index := strings.IndexByte(reference, '@'); index != -1 {
+			// `@` implies *digest*, so drop the *tag* (irrespective of what it is).
+			ref.Reference = reference[index+1:]
+			err = ref.ValidateReferenceAsDigest()
+		} else if index = strings.IndexByte(reference, ':'); index != -1 {
+			// no `@` but `:` again implies that it can only (legally) be a *digest*
+			err = ref.ValidateReferenceAsDigest()
+		} else {
+			// finally, the only possibility now is that it's a *tag*
+			err = ref.ValidateReferenceAsTag()
+		}
```

Comment:
Can be simplified as follows.
```suggestion
		} else {
			err = ref.ValidateReference()
		}
```

The `len(ref.Reference) == 0` case will be caught be line 230.

Reply:
Nice!